pipelines/batch/src/main/java/com/google/fhir/analytics/ParquetMerger.java

```diff
@@ -56,13 +70,107 @@ public class ParquetMerger {
+        pipeline
+            .apply(Create.of(getParquetFilePaths(viewName, Arrays.asList(dwh2))))
+            .apply(FileIO.matchAll())
+            .apply(FileIO.readMatches());
+
+    // The assumption here is that the schema of the old parquet files is same as the current
+    // schema, otherwise reading of the older records fail even if the new Schema is just an
+    // extension. In general, if the schema changes due to addition of new extensions, then the
+    // merging process fail. It is recommended to recreate the entire parquet files using the new
+    // schema again using the batch run.
+    PCollection<GenericRecord> records1 = inputFiles1.apply(ParquetIO.readFiles(schema));
+    PCollection<GenericRecord> records2 = inputFiles2.apply(ParquetIO.readFiles(schema));
+
+    String timestamp1 = dwh1.readTimestampFile(DwhFiles.TIMESTAMP_FILE_START).toString();
+    String timestamp2 = dwh2.readTimestampFile(DwhFiles.TIMESTAMP_FILE_START).toString();
+
+    PCollection<KV<String, GenericRecord>> dwhIdGroup1 =
+        records1
+            .apply(
+                ParDo.of(
+                    new DoFn<GenericRecord, KV<String, GenericRecord>>() {
+                      @ProcessElement
+                      public void processElement(
+                          @Element GenericRecord record,
+                          OutputReceiver<KV<String, GenericRecord>> out) {
+                        String id = record.get(ID_KEY).toString();
+                        if (id == null) {
+                          throw new IllegalArgumentException(
+                              String.format("No %s key found in %s", ID_KEY, record));
+                        }
+                        out.output(KV.of(id, record));
+                      }
+                    }))
+            .setCoder(KvCoder.of(StringUtf8Coder.of(), AvroCoder.of(schema)));
+
+    PCollection<KV<String, GenericRecord>> dwhIdGroup2 =
+        records2
+            .apply(
+                ParDo.of(
+                    new DoFn<GenericRecord, KV<String, GenericRecord>>() {
+                      @ProcessElement
+                      public void processElement(
+                          @Element GenericRecord record,
+                          OutputReceiver<KV<String, GenericRecord>> out) {
+                        String id = record.get(ID_KEY).toString();
+                        if (id == null) {
+                          throw new IllegalArgumentException(
+                              String.format("No %s key found in %s", ID_KEY, record));
+                        }
+                        out.output(KV.of(id, record));
+                      }
+                    }))
+            .setCoder(KvCoder.of(StringUtf8Coder.of(), AvroCoder.of(schema)));
+
+    PCollection<KV<String, CoGbkResult>> results;
+    if (timestamp1.compareTo(timestamp2) < 0) {
+      results =
+          KeyedPCollectionTuple.of(oldDwh, dwhIdGroup2)
+              .and(newDwh, dwhIdGroup1)
```

Comment:
Is this order correct? I mean if `timestamp1` is lexicographically smaller than `timestamp2`, then shouldn't `newDwh` be `dwhIdGroup2`?

Reply:
Used the  `Instant` class "isAfter()" method to clear confusion.